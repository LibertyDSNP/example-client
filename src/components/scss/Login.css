.Login__block {
  display: flex;
}

/*# sourceMappingURL=Login.css.map */
