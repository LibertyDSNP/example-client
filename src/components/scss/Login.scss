.Login__block {
  display: flex;
}

.Login__walletIcon {
  height: 32px;
}
