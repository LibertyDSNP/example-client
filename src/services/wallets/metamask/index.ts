import * as metamask from "./metamask";
import { Wallet } from "../wallet";

const metamaskWallet: Wallet = {
  login: async () => {
    return metamask.getWalletAddress();
  },
  logout: () => {
    return;
  },
  getAddress: async () => {
    return metamask.getWalletAddress();
  },
  getWeb3: () => {
    return metamask.getWeb3();
  },
};

export default metamaskWallet;
