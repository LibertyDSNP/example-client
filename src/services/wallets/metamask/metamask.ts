import Web3 from "web3";
import { HexString } from "../../../utilities/types";
import ethereum from "./ethereum";

export const mmweb3 = new Web3(ethereum as any);

export const isInstalled = (): boolean => {
  return Boolean(ethereum && ethereum.isMetaMask);
};

export const getWalletAddress = async (): Promise<HexString> => {
  const response = await ethereum?.request({ method: "eth_requestAccounts" });
  if (!response) throw new Error("Metamask not installed");
  return response.result ? response.result[0] : response[0];
};

export const getBalance = (account: HexString): Promise<string> => {
  return mmweb3.eth.getBalance(account).then((balance) => {
    return balance;
  });
};
