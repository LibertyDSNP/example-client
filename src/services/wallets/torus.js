import web3Torus from "./tweb3";

export const enableTorus = async (buildEnv) => {
  await web3Torus.initialize(buildEnv || "Testing");
};

export const isInitialized = () => {
  return web3Torus.initialized;
};

export const getUserInfo = async () => {
  const userInfo = await web3Torus.torus.getUserInfo();
  return userInfo;
};

export const logout = () => {
  web3Torus.torus.cleanUp().then(() => {
    sessionStorage.setItem("pageUsingTorus", false);
  });
};

export const getPublicAddress = async (verifier, verifierId) => {
  return await web3Torus.torus.getPublicAddress({
    verifier,
    verifierId,
    isExtended: true,
  });
};

export const getWalletAddress = async () => {
  const { verifier, verifierId } = await getUserInfo();
  return (await getPublicAddress(verifier, verifierId)).address;
};

export const getAccount = async () => {
  return await web3Torus.torus.accounts;
};

export const getBalance = (account) => {
  return web3Torus.web3.eth.getBalance(account).then((balance) => {
    return balance;
  });
};
