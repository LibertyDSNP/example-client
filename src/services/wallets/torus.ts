import { HexString } from "../../utilities/types";
import web3Torus, { BuildEnvironment } from "./tweb3";

export const enableTorus = async (
  buildEnv?: BuildEnvironment
): Promise<void> => {
  await web3Torus.initialize(buildEnv || "testing");
};

export const isInitialized = (): boolean => {
  return web3Torus.initialized;
};

interface UserInfo {
  email: string;
  name: string;
  profileImage: string;
  verifier: string;
  verifierId: string;
}

export const getUserInfo = async (): Promise<UserInfo> => {
  const userInfo = await web3Torus.torus?.getUserInfo(
    "This site is requesting your information"
  );
  if (!userInfo) throw new Error("Unable to find user's info");
  return userInfo;
};

export const logout = (): void => {
  web3Torus.torus?.cleanUp().then(() => {
    sessionStorage.setItem("pageUsingTorus", "false");
  });
};

type VerifierTypes = "google" | "reddit" | "discord";
type AddressType = undefined | string | { address: string };
export const getPublicAddress = async (
  verifier: VerifierTypes,
  verifierId: string
): Promise<AddressType> => {
  return await web3Torus.torus?.getPublicAddress({
    verifier,
    verifierId,
    isExtended: true,
  });
};

export const getWalletAddress = async (): Promise<string> => {
  const { verifier, verifierId } = await getUserInfo();
  const walletObject = await getPublicAddress(
    verifier as VerifierTypes,
    verifierId
  );
  if (!walletObject) throw new Error("No Wallet Address found for User");
  else if (typeof walletObject === "string") return walletObject;
  else return walletObject.address;
};

export const getBalance = async (walletAddress: HexString): Promise<string> => {
  return await web3Torus.web3.eth.getBalance(walletAddress).then((balance) => {
    return balance;
  });
};
