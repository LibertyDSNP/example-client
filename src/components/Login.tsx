import React from "react";
import { Alert, Badge, Button } from "antd";
import { WalletOutlined } from "@ant-design/icons";
import { useAppDispatch, useAppSelector } from "../redux/hooks";
import { userLogin, userLogout } from "../redux/slices/userSlice";
import * as sdk from "../services/sdk";
import * as wallet from "../services/wallets/wallet";
import * as session from "../services/session";
import LoginButton from "./LoginButton";
import Register from "./Register";

interface LoginProps {
  loginWalletOptions: wallet.WalletType;
}

const Login = ({ loginWalletOptions }: LoginProps): JSX.Element => {
  const [loading, startLoading] = React.useState<boolean>(false);
  const [alertError, setAlertError] = React.useState<string>("");
  const [popoverVisible, setPopoverVisible] = React.useState<boolean>(false);
  const [registrationVisible, setRegistrationVisible] = React.useState<boolean>(
    false
  );

  const [walletAddress, setWalletAddress] = React.useState<string>("");
  const [walletType, setWalletType] = React.useState<wallet.WalletType>(
    loginWalletOptions || wallet.WalletType.NONE
  );

  const dispatch = useAppDispatch();
  const userId = useAppSelector((state) => state.user.id);

  const setLoginAndSession = (fromId: string) => {
    setRegistrationVisible(false);
    dispatch(userLogin({ id: fromId, walletType }));
    session.saveSession({ id: fromId, walletType });
  };

  const resetLoginAndSession = (e: Error) => {
    setRegistrationVisible(false);
    setAlertError(e.message);
    setWalletAddress("");
    setWalletType(wallet.WalletType.NONE);
    session.saveSession({ id: undefined, walletType });
  };

  const login = async (walletType: wallet.WalletType) => {
    if (loading) return;
    startLoading(true);
    setWalletType(walletType);
    try {
      const walletAddrLocal = await wallet.wallet(walletType).login();
      sdk.setupProvider(walletType);
      const fromId = await sdk.getSocialIdentity(walletAddrLocal);
      if (fromId) {
        setLoginAndSession(fromId);
      } else {
        setWalletAddress(walletAddrLocal);
        setRegistrationVisible(true);
      }
    } catch (error) {
      resetLoginAndSession(error);
    } finally {
      setPopoverVisible(false);
      startLoading(false);
    }
  };

  const logout = () => {
    session.clearSession();
    if (walletType !== wallet.WalletType.NONE)
      wallet.wallet(walletType).logout();
    dispatch(userLogout());
  };
  return (
    <div className="Login__block">
      {alertError && (
        <Alert
          className="Login__alert"
          type="error"
          message={alertError}
          banner
          closable={true}
          onClose={() => setAlertError("")}
        />
      )}
      {!userId ? (
        <div>
          <LoginButton
            popoverVisible={popoverVisible}
            setPopoverVisible={setPopoverVisible}
            loginWalletOptions={loginWalletOptions}
            loading={loading}
            loginWithWalletType={login}
          />
          {registrationVisible && (
            <Register
              walletAddress={walletAddress}
              onSuccess={setLoginAndSession}
              onFailure={resetLoginAndSession}
            />
          )}
        </div>
      ) : (
        <>
          <Badge
            count={<WalletOutlined style={{ color: "#52C41A" }} />}
            offset={[-48, 8]}
          >
            <img
              className="Login__walletIcon"
              src={wallet.wallet(walletType).icon}
              alt="Wallet Symbol"
            />
          </Badge>

          <Button
            className="Login__logOutButton"
            aria-label="Logout"
            onClick={logout}
          >
            Log Out
          </Button>
        </>
      )}
    </div>
  );
};

export default Login;
