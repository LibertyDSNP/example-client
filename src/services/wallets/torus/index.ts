import * as torus from "./torus";
import { Wallet } from "../wallet";

const torusWallet: Wallet = {
  login: async () => {
    await torus.enableTorus();
    return await torus.getWalletAddress();
  },
  logout: () => {
    if (torus.isInitialized()) {
      torus.logout();
    }
  },
  getAddress: async () => {
    return await torus.getWalletAddress();
  },
  getWeb3: () => {
    return torus.getWeb3();
  },
};

export default torusWallet;
