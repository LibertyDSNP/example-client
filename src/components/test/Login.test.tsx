import Login from "../Login";
import { mount } from "enzyme";
import * as torus from "../../services/wallets/torus/torus";
import * as wallet from "../../services/wallets/wallet";
import {
  forcePromiseResolve,
  componentWithStore,
  createMockStore,
} from "../../test/testhelpers";
import { getPrefabProfile } from "../../test/testProfiles";
import * as metamask from "../../services/wallets/metamask/metamask";

jest.spyOn(torus, "logout").mockImplementation(jest.fn);
jest.spyOn(torus, "isInitialized").mockReturnValue(true);
jest
  .spyOn(torus, "enableTorus")
  .mockImplementation(jest.fn((_build) => Promise.resolve()));
jest.spyOn(metamask, "isInstalled").mockReturnValue(true);
jest
  .spyOn(metamask, "getWalletAddress")
  .mockImplementation(() => Promise.resolve("0x123"));

describe("Login Component", () => {
  describe("is logged out", () => {
    const initialState = { user: {} };
    const store = createMockStore(initialState);
    it("renders without crashing", () => {
      expect(() => {
        mount(
          componentWithStore(Login, store, {
            loginWalletOptions: wallet.WalletType.NONE,
          })
        );
      }).not.toThrow();
    });

    describe("header button triggers login sequence", () => {
      it("header button -> torus login", async () => {
        const component = mount(
          componentWithStore(Login, store, {
            loginWalletOptions: wallet.WalletType.NONE,
          })
        );
        component.find(".LoginButton__loginButton").first().simulate("click");
        component.find(".LoginButton__loginTorus").first().simulate("click");
        await forcePromiseResolve();
        expect(torus.enableTorus).toHaveBeenCalled();
      });

      it("header button -> metamask login", async () => {
        const component = mount(
          componentWithStore(Login, store, {
            loginWalletOptions: wallet.WalletType.NONE,
          })
        );
        component.find(".LoginButton__loginButton").first().simulate("click");
        component.find(".LoginButton__loginMetamask").first().simulate("click");
        await forcePromiseResolve();
        expect(metamask.getWalletAddress).toHaveBeenCalled();
      });
    });

    describe("login from login guide", () => {
      it("metamask login guide", () => {
        const initialState = { user: {} };
        const store = createMockStore(initialState);
        const component = mount(
          componentWithStore(Login, store, {
            loginWalletOptions: wallet.WalletType.METAMASK,
          })
        );
        component
          .find(".LoginButton__loginButton--quickStart")
          .first()
          .simulate("click");
        expect(metamask.getWalletAddress).toHaveBeenCalled();
      });

      it("torus login guide", () => {
        const initialState = { user: {} };
        const store = createMockStore(initialState);
        const component = mount(
          componentWithStore(Login, store, {
            loginWalletOptions: wallet.WalletType.TORUS,
          })
        );
        expect(component).toMatchSnapshot();
        component
          .find(".LoginButton__loginButton--quickStart")
          .first()
          .simulate("click");
        expect(torus.enableTorus).toHaveBeenCalled();
      });
    });
  });

  describe("is logged in", () => {
    const profile = getPrefabProfile(0);
    const walletType = wallet.WalletType.TORUS;
    const initialState = { user: { profile, walletType } };
    const store = createMockStore(initialState);

    it("renders without crashing", () => {
      expect(() => {
        mount(
          componentWithStore(Login, store, {
            loginWalletOptions: wallet.WalletType.NONE,
          })
        );
      }).not.toThrow();
    });

    it("button triggers logout sequence", () => {
      const component = mount(
        componentWithStore(Login, store, {
          loginWalletOptions: wallet.WalletType.NONE,
        })
      );
      component.find("Button").simulate("click");
      expect(torus.logout).toHaveBeenCalled();
    });
  });
});
