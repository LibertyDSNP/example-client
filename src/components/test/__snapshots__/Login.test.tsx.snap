// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component is logged out login from login guide torus login guide 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Login
        loginWalletOptions="TORUS"
      >
        <div
          className="Login__block"
        >
          <LoginButton
            loading={false}
            loginWalletOptions="TORUS"
            onButtonClick={[Function]}
            popoverVisible={false}
            setPopoverVisible={[Function]}
          >
            <Button
              aria-label="Login"
              className="LoginButton__loginButton LoginButton__loginButton--quickStart"
              onClick={[Function]}
            >
              <Wave>
                <button
                  aria-label="Login"
                  className="ant-btn LoginButton__loginButton LoginButton__loginButton--quickStart"
                  onClick={[Function]}
                  type="button"
                >
                  <LoadingIcon
                    existIcon={false}
                    loading={false}
                    prefixCls="ant-btn"
                  >
                    <CSSMotion
                      motionName="ant-btn-loading-icon-motion"
                      onAppearActive={[Function]}
                      onAppearStart={[Function]}
                      onEnterActive={[Function]}
                      onEnterStart={[Function]}
                      onLeaveActive={[Function]}
                      onLeaveStart={[Function]}
                      removeOnLeave={true}
                      visible={false}
                    >
                      <DomWrapper />
                    </CSSMotion>
                  </LoadingIcon>
                  <span
                    key=".0"
                  >
                    Log In With TORUS
                  </span>
                </button>
              </Wave>
            </Button>
          </LoginButton>
        </div>
      </Login>
    </Router>
  </BrowserRouter>
</Provider>
`;
